### PR TITLE
Long Lived XP

### PR DESCRIPTION
**Pull request by zumbak04**
## Changelog:
- Long-lived races will experience the Knowledge Overflow slowing down their Lifestyle Experience with each Perk unlocked.
- Added a game rule to control the Knowledge Overflow.

## Developer changelog:
- Added a new on-action, on_perk_gained.

## Tests:
- [x] There are no errors in `wc` files in `Documents\Paradox Interactive\Crusader Kings III\logs\error.log` except `portrait_decals.cpp:101`
- [x] The mod takes less than 5.5 GB in the Task Manager (Windows)

# How to test:
Use the [Console Commands](https://ck3.paradoxwikis.com/Console_commands) to get a Lifestyle Experience. Test with different races and tell your opinion.

### CODE REVIEW

**Comment by robmart** [comment]
A game rule for those that just want to spam perks would be nice. 
Was thinking 3 settings 
* Enabled
* Disabled for players
* Disabled for all

**Comment by zumbak04** [comment]
> A game rule for those that just want to spam perks would be nice. Was thinking 3 settings

Added in 2e64b292c657a77e0299c885bb4fe782d22ab1a1.

**Comment by zumbak04** [comment]
> They are both significantly clearer! One thing stands out for the tooltip, the number at the bottom shows the slow-down per point of Overflow, but the wording seems to imply it shows the total. Maybe change it to 'Monthly Lifestyle Experience: -5% per stack' ?

Added in e818629875c01cda9f20e84870abcc332863f863.

**Comment by Melnerag** [comment]
Tested the 'per stack' change & new settings briefly.

**'Enabled'** works, no weirdness observed.

**'AI Only'** also works, and I've tested succession where the heir is landed and has Overflow. It is properly cleared on succession. However, Switching character doesn't work - the character you played originally will count as player and not have Overflow, and the character you switch to will still count as AI. I've waited a month after the change to see if it would switch over, it doesn't.

**'Disabled'** works as well. No Overflow spotted.

Per stack text to tooltip does it's job.

**Review by Melnerag — APPROVED** [review]
Approved

**Review by TheGroxEmpire — COMMENTED** [review]
- The knowledge overflow seems to be updated after you spend a perks. This cause a huge exploit where you could hoard infinite amount of perks point without triggering knowledge overflow. Then you would be able to spend them all.
- For some reason a lot of races were having the same knowledge overflow value on tooltip as human, this was all done through console commands:
<details>

image
image
image

Notice how it has kicked in at 13 yet the tooltip said 27 perks:
image


</details>

**Comment by zumbak04** [comment]
> * For some reason a lot of races were having the same knowledge overflow value on tooltip as human, this was all done through console commands:

They all have the same Life Expectancy.

**Comment by zumbak04** [comment]
> * The knowledge overflow seems to be updated after you spend a perks. This cause a huge exploit where you could hoard infinite amount of perks point without triggering knowledge overflow. Then you would be able to spend them all.

Fixed in 363ea8978bbcc7d10ea36d4371da7ef9e27b926e.

**Review by TheGroxEmpire — COMMENTED** [review]
> Fixed in 363ea89.

Knowledge overflow update trigger seems to happen pretty random and takes few months or even sometimes years to update if you are not spending perks. I don't know exactly what triggers it but it seemed that adding a nickname was one of the triggers. Would it be possible to have the trigger as soon a perk point is earned?

File: events/lifestyles/wc_lifestyle_events.txt [hunk 1]
```diff
@@ -0,0 +1,31 @@
+﻿#Events for Mystics
```
Comment (MemeWizard) [inline events/lifestyles/wc_lifestyle_events.txt:1]: wdym "Events for Mystics" ?

File: localization/english/wc_game_concepts_l_english.yml [hunk 2]
```diff
@@ -63,4 +63,11 @@
...
\ No newline at end of file
+ game_concept_eldritch_government_desc:0 "A [dark_government|E] available to [GetTrait('being_void').GetName( GetNullCharacter )] with $doctrine_shadow_magic_name$ accepted.\nHaving high [dark_authority|E], they get all [skills|E] per [stress_level|E] and increase [county_development_growth|E]."
+ 
+ game_concept_knowledge_overflow:0 "Knowledge Overflow"
+ game_concept_knowledge_overflow_common_desc:0 "When you unlock new [perks|E] the character may feel $game_concept_knowledge_overflow$: [lifestyle_experience|E] slow downs per each [perk|E].\n\nThis pace depends on [life_expectancy|E] of [wc_race|E]. Long-lived [wc_races|E] slow down slower, but sooner. Short-lived slow down faster, but later."
+ game_concept_knowledge_overflow_desc:0 "$game_concept_knowledge_overflow_common_desc$\n\nFor example, [GetTrait('creature_human').GetName( GetNullCharacter )] characters slow down by #N [GuiScope.SetRoot(GetPlayer.MakeScope).ScriptValue('long_lived_lifestyle_xp_slowdown_human_per_perk_percent_value')|2]%#! starting after [GuiScope.SetRoot(GetPlayer.MakeScope).ScriptValue('long_lived_lifestyle_xp_slowdown_human_offset_value')|V0] [perk|E]."
```
Comment (MemeWizard) [inline localization/english/wc_game_concepts_l_english.yml:70]:
```suggestion
 game_concept_knowledge_overflow_desc:0 "$game_concept_knowledge_overflow_common_desc$\n\nFor example, [GetTrait('creature_human').GetName( GetNullCharacter )] characters slow down by #N [GuiScope.SetRoot(GetPlayer.MakeScope).ScriptValue('long_lived_lifestyle_xp_slowdown_human_per_perk_percent_value')|2]%#! per each [perk|E] starting after [GuiScope.SetRoot(GetPlayer.MakeScope).ScriptValue('long_lived_lifestyle_xp_slowdown_human_offset_value')|V0] [perk|E]."
```

File: localization/english/modifiers/wc_modifiers_l_english.yml [hunk 3]
```diff
@@ -52,4 +52,7 @@
  closed_from_another_side:0 "Closed from Another Side"
  closed_from_another_side_desc:0 "The Dark Portal is closed from another side."
  
- game_start_health:0 "Long-lived"
\ No newline at end of file
+ game_start_health:0 "Long-lived"
+ 
+ long_lived_lifestyle_xp_slowdown:0 "[knowledge_overflow|E]"
+ long_lived_lifestyle_xp_slowdown_custom_desc:0 "$game_concept_knowledge_overflow_common_desc$\n\nDue to [ROOT.Char.GetShortUINamePossessive] [life_expectancy|E], [lifestyle_experience|E] slow downs by #N [GuiScope.SetRoot(ROOT.Char.MakeScope).ScriptValue('long_lived_lifestyle_xp_slowdown_per_perk_percent_value')|2]%#! starting after [GuiScope.SetRoot(ROOT.Char.MakeScope).ScriptValue('long_lived_lifestyle_xp_slowdown_offset_value')|V0] [perk|E].\n\nPer stack:"
```
Comment (MemeWizard) [inline localization/english/modifiers/wc_modifiers_l_english.yml:58]:
```suggestion
 long_lived_lifestyle_xp_slowdown_custom_desc:0 "$game_concept_knowledge_overflow_common_desc$\n\nDue to [ROOT.Char.GetShortUINamePossessive] [life_expectancy|E], [lifestyle_experience|E] slow downs by #N [GuiScope.SetRoot(ROOT.Char.MakeScope).ScriptValue('long_lived_lifestyle_xp_slowdown_per_perk_percent_value')|2]%#! per each [perk|E] starting after [GuiScope.SetRoot(ROOT.Char.MakeScope).ScriptValue('long_lived_lifestyle_xp_slowdown_offset_value')|V0] [perk|E].\n\nPer stack:"
```

**Review by MemeWizard — APPROVED** [review]
You still can recieve exp from lifestyle events, but I think it's okay

**Comment by zumbak04** [comment]
> Knowledge overflow update trigger seems to happen pretty random and takes few months or even sometimes years to update if you are not spending perks. I don't know exactly what triggers it but it seemed that adding a nickname was one of the triggers. Would it be possible to have the trigger as soon a perk point is earned?

No, its not possible. It happens every 3 years.

**Review by TheGroxEmpire — APPROVED** [review]
That's all considered then.

File: events/lifestyles/wc_lifestyle_events.txt (hunk 1 above)
Comment (zumbak04) [reply events/lifestyles/wc_lifestyle_events.txt]: Cleared in 8da51cc1803a3b0c29c377cbe76c316ba9013218.

**Comment by zumbak04** [comment]
> You still can recieve exp from lifestyle events, but I think it's okay

I know. It shouldn't be a lot so I keep it for now.
